### PR TITLE
v0.1.0, Redesign the core of Casbin.js, support ABAC and other modes.

### PR DESCRIPTION
**Pull request by kingiw**
Signed-off-by: kingiw <kingiw@hotmail.com>

Redesign the core of Casbin.js, support for diverse mode of casbin.

Temporarily we use Node-casbin as the enforcer in Casbin.js. Coming up next is to implement a simple Enforcer in order to avoid cumbersome dependencies.

Because the core is dramatically changed, I moved the version to v0.1.0.

Signed-off-by: kingiw <kingiw@hotmail.com>

### CODE REVIEW

**Comment by kingiw** [comment]
@hsluoyz Plz review.

**Comment by hsluoyz** [comment]
@nodece please review.

File: src/Authorizer.ts [hunk 1]
```diff
@@ -59,61 +64,95 @@ export class Authorizer {
...
+        if (this.permission === undefined) {
+            this.permission = new Permission();
+        }
         this.permission.load(permission);
     }
+    
+    public async initEnforcer(s: string): Promise<void> {
+        const obj = JSON.parse(s);
+        if (!('m' in obj)) {
+            console.log("Error: No model when init enforcer.");
```
Comment (nodece) [inline src/Authorizer.ts:85]: There should be `throw` error when doesn't have model.

File: src/Authorizer.ts [hunk 2]
```diff
@@ -59,61 +64,95 @@ export class Authorizer {
...
+        }
+        const m = casbin.newModelFromString(obj['m']);
+        this.enforcer = await casbin.newEnforcer(m);
+        if ('p' in obj) {
+            for (const sArray of obj['p']) {
+                this.enforcer.addPolicy(sArray[1].trim(), sArray[2].trim(), sArray[3].trim());
```
Comment (nodece) [inline src/Authorizer.ts:91]: miss `await`.

File: src/Authorizer.ts [hunk 3]
```diff
@@ -59,61 +64,95 @@ export class Authorizer {
...
-                await this.syncUserPermission();
-                Cache.saveToLocalStorage(user, this.permission.getPermissionString(), this.cacheExpiredTime);
+            const cachedConfig = Cache.loadFromLocalStorage(user);
+            if (cachedConfig === null) {
+                const config = await this.getEnforcerDataFromSvr();
+                await this.initEnforcer(config);
+                Cache.saveToLocalStorage(user, config, this.cacheExpiredTime);
             } else {
-                this.permission.load(permStr);
+                this.initEnforcer(cachedConfig);
```
Comment (nodece) [inline src/Authorizer.ts:120]: miss `await`.

File: src/Authorizer.ts [hunk 4]
```diff
@@ -59,61 +64,95 @@ export class Authorizer {
...
 
-    public can(action: string, object: string): boolean {
-        return this.permission.check(action, object);
+    public async can(action: string, object: string): Promise<boolean> {
+        if (this.mode == "manual") {
+            return this.permission !== undefined && this.permission.check(action, object);
+        } else if (this.mode == "auto") {
+            if (this.enforcer === undefined) {
+                throw Error("Enforcer not initialized");
+                return false;
```
Comment (nodece) [inline src/Authorizer.ts:131]: Unreachable code.

File: src/Enforcer.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import * as casbin from "casbin";
+
+export default class Enforcer {
+    private e!: casbin.Enforcer;
+
+    public async init(modelStr: string, policies: string[][]) {
+        const m = casbin.newModelFromString(modelStr);
+        this.e = await casbin.newEnforcer(m);
+        this.e.addPolicies(policies);
```
Comment (nodece) [inline src/Enforcer.ts:9]: miss `await`, or returns `this.e.addPolicies(policies);`.

File: src/__test__/enforcer.test.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+[matchers]
+m = r.sub == p.sub && r.obj == p.obj && r.act == p.act
+`
+
+test('Load casbin from strings.', async () => {
+    const m = casbin.newModelFromString(s);
+    const e = await casbin.newEnforcer(m);
+    
+    e.addPolicy("alice", "data1", "read");
```
Comment (nodece) [inline src/__test__/enforcer.test.ts:40]: miss `await`.

**Review by nodece — APPROVED** [review]

**Pull request merged**